### ISSUE TITLE
NodeRed injects old nodes into .config.json at runtime

### ISSUE THREAD

**Issue by computerFriend** [node-red/node-red#1756]
This happens mostly with custom node development - testing locally & through npm.

### Steps To Reproduce:
- npm install a node locally (e.g., npm install /path/to/node/package),
- run node-red locally
- verify that the module appears in the palette & functions fine)
- stop node-red
- npm uninstall that same module
- clear browser cache & history
- search entire node red project for the module (Ctrl+F) to ensure it is completely gone from the project
- restart node-red
- See the ghost module is still there in the palette, but not functional
- stop node-red
- Reinstall the same node module from before
- start node-red
- observe that the module remains unfunctional, even though it worked in an initial install

### What happens?
- After _completely removing the node package_, this "ghost node" will be re-added to the .config.json file when Node-Red restarts, even if the browser cache was cleared.  The node will also appear in the NodeRed palette, with the same description in the info tab.

<img width="1527" alt="screen shot 2018-06-06 at 11 31 09 am" src="https://user-images.githubusercontent.com/25398257/41048717-5d0acd5e-697d-11e8-94ac-dc862d5dbb16.png">

However, if you try to add that Node to a flow, the info tab will go blank, and you can find the following error in the console (Inspect Element --> console): 

> RED.events.emit error: [nodes:add] TypeError: Cannot read property 'module' of undefined
> red.min.js:16 TypeError: Cannot read property 'module' of undefined
>     at Array.<anonymous> (red.min.js:16)
>     at Object.emit (red.min.js:16)
>     at Object.X [as add] (red.min.js:16)
>     at HTMLDivElement.drop (red.min.js:16)
>     at e.(:443/anonymous function).(anonymous function)._trigger (http://localhost:443/vendor/vendor.js:16:10003)
>     at e.(:443/anonymous function).(anonymous function)._drop (http://localhost:443/vendor/vendor.js:16:40380)
>     at e.(:443/anonymous function).(anonymous function)._drop (http://localhost:443/vendor/vendor.js:16:5029)
>     at e.(:443/anonymous function).(anonymous function).<anonymous> (http://localhost:443/vendor/vendor.js:16:42279)
>     at Function.each (vendor.js:2)
>     at Object.drop (vendor.js:16)

<img width="1527" alt="screen shot 2018-06-06 at 11 31 42 am" src="https://user-images.githubusercontent.com/25398257/41048685-4bbc3d08-697d-11e8-86b2-ec8a512b6f18.png">

### What do you expect to happen?

When a node package is uninstalled, it should be removed completely.  It should not be added to .config.json upon startup, as this causes issues.

When you re-install a node package, you should be able to do so without issue. The issues from uninstalling should not linger, especially if you have used npm uninstall & double-checked the package.json & package-lock.json files to ensure the module is gone from the first install.

### Environment:
- [ ] Node-RED version: v0.18.7
- [ ] node.js version: v9.11.1
- [ ] npm version: 6.1.0
- [ ] Platform/OS: MacBook Pro: MacOS High Sierra Version 10.13.4 (17E202)
- [ ] Browser: Google Chrome Version 66.0.3359.181 

### Comments:
Clearly there is some sort of caching happening in the background -- unfortunately, I can't find where.  I don't know enough about how the .config.json is updated at runtime. That's the missing piece for me. Please help!

**Comment by knolleary** [comment]
Hi,

Thanks for the issue report - that does sound odd.

Node-RED only adds things to .config.json that it has found as part of its scan for nodes. There is no other place it stores information about nodes, so after a restart, the only reason it would add something back into .config.json is because it has found it on disk.

The entry in .config.json includes where on disk the module was found. Can you check to see if the path it reports exists and if it exists whether it was actually uninstalled as you expected?

There is _one_ edge case in this area - If, rather than remove the whole module, you just remove one set of node files from within the module, then it might get confused. But that doesn't fit with the description you've provided.

**Comment by computerFriend** [comment]
Hi Nick,

Thanks for the quick feedback - and good cool.  I checked the path referenced in the .config.json file (which points to "~/orchestrator-core-prototype/node_modules/node-red-contrib-c8y-events/c8y-events.js", the node_modules for the local project), and I found that the module was still (partially) there, despite being "uninstalled" -- So I went & forcefully removed the folder from node_modules; then restarted node-red, & the .config.json did not self-update.  Further, I was able to re-install the module & observed fully functional behavior.

 This may indicate an issue with `npm uninstall` - who knows.  

Either way, not a Node-Red problem, & I consider this issue closed.  Very good to know how more about how the .config.json file works. Thanks again for the help!

Best,

Allie
